confmap/confmap.go

```diff
@@ -239,6 +240,23 @@ func unmarshalerHookFunc(result interface{}) mapstructure.DecodeHookFuncValue {
 	}
 }
 
+// mapstructure library doesn't override full slice during unmarshalling.
+// Origin issue: https://github.com/mitchellh/mapstructure/issues/74#issuecomment-279886492
+// To address this we zeroes every slice before unmarshalling unless user provided slice is nil.
+func zeroesSliceHookFunc() mapstructure.DecodeHookFuncValue {
+	return func(from reflect.Value, to reflect.Value) (interface{}, error) {
+		if from.Kind() == reflect.Slice && from.IsNil() {
```

Comment:
Len() == 0? No need to re-initialize if empty?

Reply:
If user provides empty list in the config explicitly then we zeroes the default value of the field.

Reply:
@bogdandrutu do you mean if both source and destination are 0 then we should just return?

Reply:
Lets start with this:
`nil != initialized and len 0`

```yaml
foo:
```

```yaml
foo: []
```

What is the behavior we want in these cases? Please document them somewhere, in order for me to understand if the implementation is ok.

Reply:
Ping on this.